drivers/at30tse75x: add device driver for AT30TSE75x temperature sensor

daniel-k opened the pull request:
This PR add a device driver for the AT30TSE75x series temperature sensors included in the [ATIO1-XPRO](http://www.atmel.com/tools/ATIO1-XPRO.aspx) dev board. There's also a EEPROM on the same chips, for which I will add driver support later.

This PR weakly depends on #3751, but should work without error feedback nevertheless.

PeterKietzmann commented:
@daniel-k thanks for the nice driver! Is the review urgent? Generally I would review this PR, but currently this may take some time. Does someone own this piece of hardware? Maybe @kaspar030?

PeterKietzmann commented:
Ah, and I will assign myself for now. If someone else has capacities for a review, feel free to assign yourself :-) !

daniel-k commented:
No, it's not urgent. Nothing depends on it, so take your time. Maybe the EEPROM driver will be finished before this get's merged (depending on my motivation) but I'd open a separate PR anyway.

daniel-k commented:
@PeterKietzmann 

I'm not sure if it's a good idea to call `i2c_init_master()` inside the driver. Shouldn't this be called by the user code, because every i2c driver would then initialize the master device again? I'm asking because other driver are doing the same.

PeterKietzmann commented:
I know I know.. Please check #2528?

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+
+
+#include "periph/i2c.h"
+#include "hwtimer.h"
```
@daniel-k, sorry for the delay. Could you adapt the implementation to use `xtimer` please

daniel-k replied:
Done, see 0d171a8.

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+    return temp_int + (frac_multiplier * AT30TSE75X_FRACTIONAL_BASE);
+}
+
+static int at30tse75x_get_register(at30tse75x_t* dev, uint8_t reg, uint16_t* data)
```
Why `uint16_t* data` but casting `(char*) data` in line 40?

PeterKietzmann replied:
Ups, forget about it. Sorry

daniel-k replied:
Because of
```c
int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int length);
```

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+    }
+
+    switch(mode) {
+    case AT30TSE75X_MODE_ONE_SHOT:
```
I'm not sure out of my head if it's written in our coding conventions, but I think that `case` is indented against `switch` mostly

daniel-k replied:
I'm just aware of [these coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions) and they don't mention this. But I don't really have an opinion on this, so I can adapt it if you wish.

PeterKietzmann replied:
Honestly I don't even care either. Just wanted to make it look similar to other code. Are there other opinions?

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+    }
+
+    if(at30tse75x_set_config(dev, config) != 0) {
+        return -1;
```
What do you think about  returning `-2` here (and updating the doc of course)?

PeterKietzmann replied:
In addition: That question also applies to other functions

daniel-k replied:
Can you explain why returning `-2`?

PeterKietzmann replied:
My general idea was to have a more detailed error reporting. `-1` would mean "mode unavailable", `-2` would mean "error setting configuration". I just wanted to hear your opinion on this. IMO it is okay the way you did, but there might be others that want a finer reporting

daniel-k replied:
Ah okay, didn't get this completely right the first time. Yes I like that, so `-1` is `general i2c failure` and `-2` wrong user input in this case. So `-2` will be returned in the `switch::default` statement. Will update.

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+    }
+
+    switch(polarity)
+    {
```
Please put the brace in the same line than `switch`

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,312 @@
+                        AT30TSE75X_CONFIG__RESOLUTION_SHIFT;
+
+        /* Wait until conversion is finished */
+        xtimer_usleep((uint32_t)(25000 << resolution));
```
I understand that the conversion time is dependent from the resolution. But why shifting `25000`?

daniel-k replied:
The conversion time is a multiple of 25ms. See [datasheet](http://www.atmel.com/Images/Atmel-8751-DTS-AT30TSE752-754-758-Datasheet.pdf) p.47.

PeterKietzmann commented on drivers/include/at30tse75x.h:
```diff
@@ -0,0 +1,295 @@
+int at30tse75x_set_fault_tolerance(at30tse75x_t* dev, at30tse75x_fault_tolerance_t tolerance);
+
+/**
+ * @brief   Set T_Low limit
```
I know it's a bit picky but I would have liked to read "Set lower temperature limit" (or similar) here

daniel-k replied:
This was actually kind of intentional. Someone who uses these features will need to read the datasheet anyway to understand what it's doing (not obvious IMO), so I named it like it is written in the datasheet. But again no strong opinion.

PeterKietzmann replied:
Hmm, sounds reasonable. Fine, leave it as is!

PeterKietzmann commented on drivers/include/at30tse75x.h:
```diff
@@ -0,0 +1,295 @@
+int at30tse75x_set_limit_low(at30tse75x_t* dev, int8_t t_low);
+
+/**
+ * @brief   Set T_High limit
```
Also here: maybe spell in full what T_High means

PeterKietzmann commented on sys/shell/commands/shell_commands.c:
```diff
@@ -198,5 +202,8 @@ const shell_command_t _shell_command_list[] = {
 #endif
+#ifdef MODULE_AT30TSE75X
+    {"at30tse75x", "Test AT30TSE75X temperature sensor", _at30tse75x_handler},
+#endif
```
Would you mind moving these lines to the other sensor drivers? E.g. starting behind `MODULE_LTC4150`

PeterKietzmann commented on sys/shell/commands/shell_commands.c:
```diff
@@ -120,6 +120,10 @@ extern int _zep_init(int argc, char **argv);
+#ifdef MODULE_AT30TSE75X
+extern int _at30tse75x_handler(int argc, char **argv);
+#endif
+
```
Would you mind moving that some line upwards, just behind the lines for `MODULE_LTC4150`

PeterKietzmann commented on tests/driver_at30tse75x/main.c:
```diff
@@ -0,0 +1,52 @@
+    shell_run(&shell);
+
+    return 0;
+}
```
It's been a while that I cared about shell programs but is all that stuff still current? Comare e.g. [the `periph_spi` test](https://github.com/RIOT-OS/RIOT/tree/master/tests/periph_spi)

daniel-k replied:
You might be right. There were some changes in the near past that I didn't keep track of. Will adapt this tomorrow.

PeterKietzmann commented:
@daniel-k thanks again for that PR. Code looks nice overall. Please adapt the discussed changes. After that we should see what the build tests report. Again: Does anyone own that hardware (e.g. ATIO1-XPRO dev board) ? Would be nice if the PR is tested by a second party before merging it.

daniel-k commented:
@PeterKietzmann 
When testing the board again it suddenly not worked anymore. After some (too much) debugging I found out that the mcu is too fast calling driver functions that access I2C successively so that the bus fee time of at least 600ns (p.48) is violated and communication fails. I fixed this by prefixing every bus access by the driver with a 1us spin. Considering the already slow ooperation of I2C (100kHz) this shouldn't hurt too much.

Additionally I rebased on master and squashed everything that has been reviewed yet into the first commit, as well as some cleanups and a sanity check of the supplied device address.

PeterKietzmann commented on drivers/at30tse75x/at30tse75x.c:
```diff
@@ -0,0 +1,320 @@
+        return -1;
+    }
+
+#if ENABLE_DEBUG
```
I think you can remove these preprocessor instructions now

PeterKietzmann commented on drivers/include/at30tse75x.h:
```diff
@@ -0,0 +1,311 @@
+ * @return                   0 on success
+ * @return                  -1 on error
+ */
+int at30tse75x_set_resolution(at30tse75x_t* dev, at30tse75x_resolution_t resolution);
```
You need to add a note for `return -2`

PeterKietzmann commented:
@daniel-k thanks for addressing my comments. I'm fine with the small delay before the bus access. IIRC we had a similar situation for an other sensor (but with the CS line of SPI). Please address my latest minor comments. After that we can run Travis.

daniel-k commented:
Travis running, but still needs squashing.

PeterKietzmann commented:
@daniel-k Travis failed for several boards because of a redefinition of `MODULE_AT30TSE75X` in *sc_at30tse75x.c*.  Please remove that define out of the script. AFAIK this is generated by the makefile.

PeterKietzmann commented on sys/shell/commands/sc_at30tse75x.c:
```diff
@@ -0,0 +1,160 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include "at30tse75x.h"
+#define MODULE_AT30TSE75X
```
Why did you introduce this in late before last commit? Travis complains about that define. (He complains about redefinition and I think it is initially defined by the makefile system)

daniel-k replied:
This slipped through by accident. Eclipse doesn't know those defines, so auto-completion doesn't work inside unknown `#ifdef`s.

gebart replied:
Run make eclipsesym to create an eclipse symbol xml file that you can
import into your eclipse project. There's a page in the wiki about it.
On Oct 2, 2015 12:17 PM, "Daniel Krebs" <notifications@github.com> wrote:

> In sys/shell/commands/sc_at30tse75x.c
> <https://github.com/RIOT-OS/RIOT/pull/3755#discussion_r41007887>:
>
> > + * @{
> > + *
> > + * @file
> > + * @brief       Provides shell commands to test AT30TSE75x temperature sensor
> > + *
> > + * @author      Daniel Krebs <github@daniel-krebs.net>
> > + *
> > + * @}
> > + */
> > +
> > +#include <stdio.h>
> > +#include <string.h>
> > +#include <stdlib.h>
> > +#include <stdbool.h>
> > +#include "at30tse75x.h"
> > +#define MODULE_AT30TSE75X
>
> This slipped through by accident. Eclipse doesn't know those defines, so
> auto-completion doesn't work inside unknown #ifdefs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3755/files#r41007887>.
>

PeterKietzmann commented:
Oh no, there are 14 commits :-) ! @daniel-k please squash again.  @OlegHahm the code looks fine to me and Travis succeeded. But no one else has the hardware to test. I tend to merge anyway. Ok with you?

daniel-k commented:
Finally, travis is happy :-D

PeterKietzmann commented:
As far as I see it, nobody objects merging this PR. Again: I did not test this PR with hardware but the code looks fine now and makes Travis green. Because I'm away for two weeks from now on, I will just merge the PR, so @daniel-k can use the driver in his project in a more comfortable way. Also I do trust him that he'll take care of it, just *if* it causes some problems (but I assume that won't be the case ;-) )

PeterKietzmann commented:
So ACK and go

daniel-k commented:
> Also I do trust him that he'll take care of it, just if it causes some problems

Sure! Thanks for merging!
